dbt/scripts/utils/townships.py

```diff
@@ -0,0 +1,383 @@
+class Township:
+    """Class that represents a township for the purposes of QC reporting.
+
+    The schema for this class must match the town_active_schedule.csv data file
+    stored in this directory, since the script parses that file and turns each
+    row into an instance of this class."""
+
+    township_code: str
+    township_name: str
+    tri_code: str
+    tri_name: str
+    # Dates that the town becomes "active" (i.e. eligible for QC reporting)
+    # or inactive
+    active_start_date: datetime.date | None
+    active_end_date: datetime.date | None
+
+    def is_reassessed_during(self, year: int) -> bool:
+        """Helper function to determine if a town in a given year is undergoing
+        triennial reassessment"""
+        # 2024 is the City reassessment year (tri code 1), so
+        # ((2024 - 2024) % 3) + 1 == 1, and so on for the other two tris
+        return str(((year - 2024) % 3) + 1) == self.tri_code
+
+    def is_active_on(self, date: datetime.date) -> bool:
+        """Helper function to determine if a town is 'active' (i.e. eligible
+        for QC reporting) on a given date"""
+        if self.active_start_date and self.active_end_date:
+            return self.active_start_date <= date <= self.active_end_date
+        return False
+
+
+# Get a list of townships with their code and schedule, either from a schedule
+# file if one exists or from a static definition
+TOWNSHIP_SCHEDULE_PATH = pathlib.Path("scripts/utils/town_active_schedule.csv")
+TOWNSHIPS: list[Township]
+if TOWNSHIP_SCHEDULE_PATH.is_file():
+    with open(TOWNSHIP_SCHEDULE_PATH.resolve()) as town_schedule_fobj:
+        date_format = "%m/%d/%Y"
+        TOWNSHIPS = [
+            Township(
+                township_code=town["township_code"],
+                township_name=town["township_name"],
+                tri_code=town["tri_code"],
+                tri_name=town["tri_name"],
+                active_start_date=(
+                    datetime.datetime.strptime(
+                        town["active_start_date"], date_format
+                    ).date()
+                    if town.get("active_start_date")
+                    else None
+                ),
+                active_end_date=(
+                    datetime.datetime.strptime(
+                        town["active_end_date"], date_format
+                    ).date()
+                    if town.get("active_end_date")
+                    else None
+                ),
+            )
+            for town in csv.DictReader(town_schedule_fobj)
+        ]
+else:
+    # Instantiate the list of townships with no active/inactive date data,
+    # so that the script still works without access to the town_active_schedule
+    TOWNSHIPS = [
+        Township(
+            township_code="10",
+            township_name="Barrington",
+            tri_code="2",
+            tri_name="North",
+            active_start_date=None,
+            active_end_date=None,
+        ),
```

Comment:
**suggestion (non-blocking):** Could you DRY this up a bit by just looping through a list of tuples with the positional values for each township? i.e.

```python
TOWNSHIPS_DATA = [
    ("10", "Barrington", "2", "North"),
    ("11", "Berwyn", "3", "South"),
    # blah blah blah
]

TOWNSHIPS = [
    Township(
        township_code=code,
        township_name=name,
        tri_code=tri_code,
        tri_name=tri_name,
        active_start_date=None,
        active_end_date=None,
    )
    for code, name, tri_code, tri_name in TOWNSHIPS_DATA
]
```

Reply:
Better yet, why not just have a default `town_active_schedule.csv` file with all this info but the start and end date blank, then just instantiate the townships based on what's available in that sheet.

Reply:
I've refactored the design to move away from the activity schedule, so this point is moot now.